### PR TITLE
vmware_guest: various customizations changes and fixes

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

v2.3

##### SUMMARY

Most of my queued changes were already implemented by @aperigault !

This was still open
- Typos
- Various fixes to dict.get() without quotes
- Add defaults to `fullname` and `orgname` (so they are no longer mandatory)
- ~~Add default to `joinworkgroup` to WORKGROUP (maybe not a good idea afterall ?)~~
- Add missing timezone implementation
- Remove the customize flag from the options
- Rename 'customizations' to 'customization' (cfr VMware docs and fora)

### CODE REVIEW

**Comment by dagwieers** [comment]
@nerzhul @aperigault @dav1x @nitzmahone @jhawkesworth @jctanner Feedback welcome

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:443]: oh tricky i didn't noticed this typo

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (aperigault) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: I don't understand. With your change we attempt to customize a vm created from scratch ?
I think we can't deploy VM from template without use customization with this.

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: @aperigault Good question, I thought it was possible to customize newly created VMs as well.
Is that not correct ?

Comment (nerzhul) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: this condition is correct as vm_obj is none on vm creation from scratch. This case is handled using same condition in every self.configure_xx
why customization should be forbidden ? If i want to create a VM from scratch using a PXE boot and after your sysprep ?

Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Ok we can use sysprep with from scratch VM. But if we run customization in all cases why this condition ?
I think it should be : if we set customizations params : if len(self.params['customize']) > 0:

Comment (nerzhul) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: in fact, you are wrong @aperigault because if you add that you will break the guest_map customizations which are handled in customize_vm and doesn't require customize variable namespace

Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Oh yes! I missed it! But there always is a problem. How can I simply disable customization ? We should add tests in customize_vm or split customize_vm in two function ?

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Well, I was wondering the same thing. It only affects the stuff that defaults to something.
Does it really matter ? Except maybe speed ?
I don't mind re-introducing `customize:`, default it to `customize: yes`, although I don't like the name.

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Or assume that an empty customizations means no customizations (like you did).
It's easy to add a single item if needed.

Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: I prefer the second one. If you want to use customization (others than network customization) you need to set customization params.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (aperigault) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: I thought about it and I would prefer to not set default value for this param. It could be better to explicitly enable joindomain or joinworkgroup.

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: I agree (as put in the description). Yesterday I thought it was a good idea, but maybe people have it already configured correctly in the template and we don't want to mess with it. Will undo.

**Comment by dagwieers** [comment]
Shippable fails on:
```
Error response from daemon: Get https://registry-1.docker.io/v2/drydock/u14pyt/manifests/prod: dial tcp 52.0.53.94:443: connect: network is unreachable
```

Maybe we need an ansibot command to collect these kind of Shippable errors for statistics/troubleshooting ?

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Ok, agreed ! And implemented !

**Comment by dagwieers** [comment]
ready_for_review

**Comment by dagwieers** [comment]
ready_for_shipit :-)

**Comment by dagwieers** [comment]
Important rename from "customizations" to "customization" (cfr. VMware documentation and fora).

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: please remove the "!"

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: Remove "a non-blank"

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
ready_for_review

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (aperigault) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: You missed to remove this line
We already talk about it but what is the goal of this part ?
(vm_obj is None or self.should_deploy_from_template())

**Comment by dagwieers** [comment]
@nerzhul I don't understand why you wanted me to remove the "non-blank" from the statement though. I test both if the password was defined and is non-empty. A non-empty password will cause the customization to fail and the module will end with a vague message that there was an error with customization.

So it is important that the password is not empty, as much as it is important that it is provided.

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (aperigault) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: I ask again: When this condition will be skipped ?
I understand: When we deploy from scratch or from template. There is an other case ?

**Comment by aperigault** [comment]
Can we merge that to fix #20038 ?

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (aperigault) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: With this we can't customize from scratch vm. I think we only need:
if len(self.params['customization']) > 0:

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Changed. I am going tired and don't even want to discuss anything anymore.

I'd rather get this pulled and then improved with individual PRs. I have rebased/merged conflicts 4 times now because fixes in this patch are being merged faster than this PR. So please merge this, and then offer PRs instead.

**Comment by dagwieers** [comment]
Please merge this one.

**Comment by aperigault** [comment]
LGTM after shippable fix

**Comment by aperigault** [comment]
@nerzhul @gundalow @jctanner Can you merge this PR ? There is a lot of important fixes. Thanks

**Comment by nerzhul** [comment]
I need to test it before accept it, wait for 1 or 2 hours

**Review by nerzhul — APPROVED** [review]
ship_it

File: lib/ansible/modules/cloud/vmware/vmware_guest.py
Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: To be honest I wasn't sure what my original code was after a merge-glitch during rebase (as it wasn't changed in my production code). So I put back what was there originally, not what I changed it to previously...

**Pull request merged**
